cpp/src/arrow/compute/kernels/vector_statistics.cc

```diff
@@ -0,0 +1,208 @@
+                           ArrayData* out, KernelContext* ctx) {
+    DCHECK_EQ(out->buffers.size(), data.buffers.size());
+    out->null_count = data.null_count.load();
+    out->length = data.length;
+    out->buffers[0] = data.buffers[0];
+    ARROW_ASSIGN_OR_RAISE(out->buffers[1], ctx->Allocate(out->length * sizeof(CType)));
+    // Avoid leaving uninitialized memory under null entries
+    std::memset(out->buffers[1]->mutable_data(), 0, out->length * sizeof(CType));
+
+    const CType* in_values = data.GetValues<CType>(1);
+    CType* out_values = out->GetMutableValues<CType>(1);
+
+    auto visit = [&](int64_t position, int64_t length) {
+      for (int64_t i = position; i < position + length; ++i) {
+        if (in_values[i] < quantiles.lower_bound) {
+          out_values[i] = quantiles.lower_bound;
+        } else if (in_values[i] > quantiles.upper_bound) {
+          out_values[i] = quantiles.upper_bound;
+        } else {
+          // NaNs also fall here
+          out_values[i] = in_values[i];
+        }
+      }
+    };
+    arrow::internal::VisitSetBitRunsVoid(data.buffers[0], data.offset, data.length,
+                                         visit);
+    return Status::OK();
+  }
+};
+
+template <typename Unused, typename Type>
+struct WinsorizeChunked {
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    return Winsorize<Unused, Type>::ExecChunked(ctx, batch, out);
+  }
+};
+
+Result<TypeHolder> ResolveWinsorizeOutput(KernelContext* ctx,
+                                          const std::vector<TypeHolder>& in_types) {
+  DCHECK_EQ(in_types.size(), 1);
+  return in_types[0];
+}
+
+const FunctionDoc winsorize_doc(
+    "Winsorize an array",
+    ("This function applies a winsorization transform to the input array\n"
+     "so as to reduce the influence of potential outliers.\n"
+     "NaNs and nulls in the input are ignored for the purpose of computing\n"
+     "the lower and upper quantiles.\n"
+     "The quantile limits can be changed in WinsorizeOptions."),
+    {"array"}, "WinsorizeOptions", /*options_required=*/true);
+
+}  // namespace
+
+void RegisterVectorStatistics(FunctionRegistry* registry) {
+  static const auto default_winsorize_options = WinsorizeOptions();
+
+  auto winsorize = std::make_shared<VectorFunction>(
+      "winsorize", Arity::Unary(), winsorize_doc, &default_winsorize_options);
+
+  VectorKernel base;
+  base.init = WinsorizeState::Init;
+  base.mem_allocation = MemAllocation::NO_PREALLOCATE;
+  base.null_handling = NullHandling::COMPUTED_NO_PREALLOCATE;
+  base.can_execute_chunkwise = false;
+  // The variable is ill-named, but since we output a ChunkedArray ourselves,
+  // the function execution logic shouldn't try to wrap it again.
+  base.output_chunked = false;
+
+  for (const auto& ty : NumericTypes()) {
+    base.signature = KernelSignature::Make({ty->id()}, &ResolveWinsorizeOutput);
+    base.exec = GenerateNumeric<Winsorize, /*Unused*/ void>(ty->id());
```

Comment:
`/*Unused=*/`?

Reply:
Well, it's not the parameter name, the comment just flags that the `Winsorize` class will ignore that parameter.